feat: Login view에서 username 입력받아 저장하고 메인화면으로 전환

SeoSiun opened the pull request:
figma 화면대로 UI 구현하고,
submit 버튼을 누르면 입력한 username을 UserDefault에 저장하고 메인화면으로 전환하며
submit later 버튼을 누르면 바로 메인화면으로 전환하게 함.

SeoSiun committed:
feat: Login UI 구현 및 입력받은 username 저장

figma화면대로 username입력받는 textField 및
submit, submit later 버튼 구현
오토레이아웃은 안하고 그냥 프레임으로 위치 설정함

Resolves: GITIT-46

SeoSiun committed:
feat: submit, submit later 버튼 누르면 main 화면으로 넘기기

storyboard에서 mainTabBarController의 id를 "main"으로 설정함.

Resolves: GITIT-46

sweethoneybee reviewed (changes requested):
## 코드 관련 피드백
1. ~viewDidLoad와 같은 override 함수들을 프로퍼티 정의 바로 다음에 작성해주는 게 가독성이 더 좋을 것 같아요. override 메소드, UI 관련 메소드, objc 메소드, IBAction, 퍼블릭메소드, 프라이빗 메소드 순 정도로 정렬하면 좋습니다.~ 다시 생각해보니 override 메소드를 젤 위에 두는 것 말고는 가독성 있게만 구성해주시면 될 것 같습니다. 저도 좋은 방법이 무엇일지 계속 고민 중이라 자꾸 말이 바뀌네요 ㅋㅋ. 
2. implicitly unwrapped optional은 지양해주세요(프로퍼티 타입 정의 시에 ! 붙이는 것). 정말 필요한 경우(지금 생각나는 것도 없지만)를 제외하고는 ! 는 지양하는 게 더 안정성 있는 앱을 구축할 수 있게 합니다.
3. 라벨, 버튼 등을 생성하는 코드의 경우에 지금처럼 함수로 구분해 놓으니 가독성이 올라갔네요. 이와 비슷한 방식으로 클로저를 활용해보는 것도 좋을 것 같아요. 
ex) 
```swift
let label: UILabel = { label in 
    label.frame = CGRect(x: 55, y: 300, width: 200, height: 40)
    label.font = label.font.withSize(23)
    label.textColor = UIColor.black
    label.text = "My username"
    return label
}(UILabel())

self.view.addSubview(label)
```
4. StoryBoardID는 ViewController의 이름을 따라가주세요. 지금 main이니, 'MainTabBar'가 ID에 더 적합합니다.
5. 코드는 한줄에 하나의 statement만 작성해주세요. 세미콜론(;)으로 한줄에 여러개의 statement를 작성하는 것보다 줄수가 조금 길어지더라도 한줄에 하나의 statement만 작성하는 게 가독성에 더 좋을 것 같아요. (77번째 줄이 현재 한 줄에 두 개가 들어가 있어요)
6. `touchUpBtnSubmmit()`, `touchUpBtnSubmmitLater()` 메소드가 오타가 난 것 같아요. m이 두번 들어갔습니다
7. ~`touchUpBtnSubmmit()`, `touchUpBtnSubmmitLater()` 에서 굳이 이벤트의 방식까지 명시해줄 필요는 없을 것 같아요. addTarget으로 메소드를 설정할 때 event가 결정되니깐요.~ -> 더 생각해보니 현재 이름 유지하는 게 더 좋을 것 같습니다. 메소드가 어떤 event에 사용될지를 이름에 명시하는 게 좋은 방식인 것 같아요
8. target-action 방식에 활용하려는 메소드는 지금 우리가 스토리보드에서 메소드를 연결하지 않기 때문에 @IBAction 어노테이션보다는 @objc 어노테이션이 지금 상황에 맞을 것 같아요. 

## 기능 구현 관련 피드백
1. 화면 전환 시, `nextVC.modalPresentationStlye`이 현재 `Automatic`으로 되어있어 화면 전환시 PageSheet 방식으로 됩니다. PageSheet는 일시적인 화면에 대해서 주로 적용을 하니 이 상황에서는 FullScreen 방식이 필요해요. 코드로 변경하셔도 되고, 아니면 제가 로그인뷰 -> 메인탭바뷰로 이동하는 세그웨이는 미리 세팅을 해놓았는데(id까지) 확인해보시고 활용하셔도 좋을 것 같아요. 편한 방식을 택해서 구현해주세요
2. textField 입력 후 입력을 마치려 해도 키보드가 내려가지 않습니다. 일반적으로 유저들은 키보드의 리턴키를 누르거나, 입력창을 제외한 뷰를 터치하면 키보드가 내려간다고 인식하니 이에 맞게 수정해주세요. 리턴키 처리는 UITextFieldDelegated의 `textFieldShoudReturn(_:)` 메소드를 구현하면 되고, 입력창 외의 터치는 UIViewController의 `touchesBegin(_:with:)` 메소드를 오버라이드해서 구현하시면 됩니다.

SeoSiun committed:
refactor: 코드 관련 피드백 반영

1. override mothod를 맨 위로 올림
2. impicitly unwrapped optional을 없앰
3. 클로저를 사용해서 라벨, 버튼 등 생성
4. StoryBoardID 수정 및 어노테이션 objc로 변경
5. 오타 수정

Resolves:GITIT-46

SeoSiun committed:
feat: 기능 관련 피드백 반영 및 버튼 활성화 기능 추가

화면 전환시 FullScreen 방식 사용
리턴키를 누르거나 입력창을 제외한 뷰 터치 시 키보드가 내려가게 함
textFieldUsername에 내용을 입력하지 않으면 submit 버튼 비활성화 상태로 설정

Resolves: GITIT-46

SeoSiun committed:
feat: username에 알파벳, 숫자, 언더바 제외 입력 거부

자동 대문자, 오타 수정 안 되게 함
알파벳, 숫자, 언더바 제외 문자의 입력을 제한

Resolves: GITIT-46

sweethoneybee reviewed (changes requested):
피드백 안 드렸었는데도 영문, 숫자, 언더바 입력제한을 잘 반영해주셨네요! 우선 시언님 커밋에 대해 리뷰하고, 윤배님과의 공통된 피드백을 드리겠습니다(거의 유사한 로직이기에).  

1. `MARK:` 보단 `MARK: -`를 사용해서 주석 달아주시면 더 보기 좋습니다.  
2. 지금 프로퍼티를 선언과 동시에 초기화하고 있는데 이 경우 메모리가 낭비될 수 있어요. 또 초기화를 했는데 또 viewDidLoad 메소드에서 객체를 생성해서 할당해주고 있으니, 사용하지 않는 객체를 할당했다가 해제하는 상황인 거예요. ?를 사용해서 옵셔널 타입으로만 선언해주고 viewDidLoad에서 할당해주세요


윤배님과 동일한 피드백은 아래와 같습니다. 아래 두 피드백은 두분이 머지 한 후에 적용한다고 생각하시면 됩니다.

1. 지금은 사용자 입장에서는 영어 소문자, 숫자, 언더바만 입력 가능하다는 걸 알 방법이 없어요. 그래서 그것에 대한 피드백이 필요해요(그 외를 입력하면 알림을 주는 등). 그렇지 않으면 앱이 제대로 동작하고 있다고 생각하지 않을 거예요. 이 문제를 어떻게 해결할지 생각해보고 Login View, Setting View에 적용해야 해요(둘 다 지금 로직이 동일하니). 우리가 웹에서 회원가입을 할 때 잘못된 입력을 넣으면 어떻게 알려주는지를 관찰해보면 답을 얻을 수 있을 거예요.  
2. 윤배님과 시언님 둘다 서로 다른 곳에서 String Extension을 구현하고 있는데 둘다 같은 메소드를 구현하고 있어요. 아마 합치면 컴파일에러가 날 것 같아요. 추후에 중복되는 부분을 분리하여 Extension 디렉토리에서 `String.hasCharacters()`를 구현해주세요

SeoSiun committed:
Merge branch 'main' into GITIT-46-saveUsername
